src/control/positionControl/pathPlanning/NumTreesPlanning.cpp

```diff
@@ -0,0 +1,214 @@
+//
+// Created by ratoone on 09-12-19.
+//
+
+#include "control/positionControl/pathPlanning/NumTreesPlanning.h"
+
+namespace rtt::ai::control{
+
+NumTreesPlanning::NumTreesPlanning(const CollisionDetector &collisionDetector) :
+    collisionDetector(collisionDetector) {}
+
+std::list<rtt::Vector2>
+NumTreesPlanning::computePath(const rtt::Vector2 &robotPosition, const rtt::Vector2 &targetPosition) {
+    DT = 0.3 / rtt::ai::GameStateManager::getCurrentGameState().getRuleSet().maxRobotVel;
+    if (DT > 0.12) DT = 0.12;
+    if (DT < 0.06) DT = 0.06;
+
+    auto path = tracePath(robotPosition, targetPosition);
+    double goToTimeInFuture = 0.4;
+    auto targetPathPoint = static_cast<unsigned long>(goToTimeInFuture/DT);
+    if (path.size() < targetPathPoint) {
+        return std::list<Vector2>(1, targetPosition);
+    }
+    return path;
+}
+
+std::list<Vector2> NumTreesPlanning::tracePath(const rtt::Vector2 &currentPosition, const rtt::Vector2 &targetPosition) {
+
+// compAStar compares the amount of collisions first (max 3 diff), then sorts the paths based on an approximation on the
+// length of path that still has to be calculated, using straight lines towards the half-way targets, and the final
+// target, while taking into account the length of path already calculated.
+    auto compAStar = [targetPosition, this](PathPointer lhs, PathPointer rhs) {
+        if (lhs->collisions - rhs->collisions > 2)
+            return true;
+        else if (lhs->collisions - rhs->collisions < - 2)
+            return false;
+        else
+            return (lhs->maxVel()*lhs->t + remainingStraightLinePathLength(lhs->pos, lhs->currentTarget, targetPosition))
+                   > (rhs->maxVel()*rhs->t
+                      + remainingStraightLinePathLength(rhs->pos, rhs->currentTarget, targetPosition));
+    };
+
+    std::vector<rtt::ai::control::PathPoint> path;
+
+    // create a new path with a root
+    std::priority_queue<PathPointer, std::vector<PathPointer>, decltype(compAStar)> pathQueue(compAStar);
+    PathPointer root = std::make_shared<rtt::ai::control::PathPoint>();
+    root->currentTarget = targetPosition;
+    root->pos = currentPosition;
+    root->vel = Vector2();
+    root->acc = Vector2();
+    root->t = 0;
+    root->collisions = 0;
+    pathQueue.push(root);
+
+    auto start = std::chrono::duration_cast< std::chrono::milliseconds >(
+            std::chrono::system_clock::now().time_since_epoch()
```

Comment:
Dont' use system_clock, use the tick from roboteam_ai world_new.

If that's not accurate enough then maybe we should consider switching those ticks to ms or ns as im p sure it's seconds rn

Reply:
This is to measure calculation time, not world time. This was already in the original code.

Reply:
So this is perferctly fine.